usr/src/uts/common/fs/zfs/dmu_traverse.c

```diff
@@ -231,18 +232,30 @@ traverse_visitbp(traverse_data_t *td, const dnode_phys_t *dnp,
 
 	if (bp->blk_birth == 0) {
 		/*
-		 * Since this block has a birth time of 0 it must be a
-		 * hole created before the SPA_FEATURE_HOLE_BIRTH
-		 * feature was enabled.  If SPA_FEATURE_HOLE_BIRTH
-		 * was enabled before the min_txg for this traveral we
-		 * know the hole must have been created before the
-		 * min_txg for this traveral, so we can skip it. If
-		 * SPA_FEATURE_HOLE_BIRTH was enabled after the min_txg
-		 * for this traveral we cannot tell if the hole was
-		 * created before or after the min_txg for this
-		 * traversal, so we cannot skip it.
+		 * Since this block has a birth time of 0 it must be one of two
+		 * things: a hole created before the SPA_FEATURE_HOLE_BIRTH
+		 * feature was enabled, or a hole which has always been a hole
+		 * in an object.  This occurs when an object is created and then
+		 * a write is performed that leaves part of the file as zeroes.
+		 * Normally we don't care about this case, but if this object
+		 * has the same object number as an object that has been
+		 * destroyed, callers of dmu_traverse may not be able to tell
+		 * that the object has been recreated.  Thus, we must give them
+		 * all of the holes with birth == 0 in any objects that may have
+		 * been recreated. Thus, we cannot skip the block if it is
+		 * possible the object has been recreated. If it isn't, then if
+		 * SPA_FEATURE_HOLE_BIRTH was enabled before the min_txg for
+		 * this traversal we know the hole must have been created before
+		 * the min_txg for this traversal, so we can skip it. If
+		 * SPA_FEATURE_HOLE_BIRTH was enabled after the min_txg for this
+		 * traversal we cannot tell if the hole was created before or
+		 * after the min_txg for this traversal, so we cannot skip it.
```

Comment:
As you can see from the code below, if the feature was enabled in the same txg as min_txg, we treat it the same as being enabled before the min_txg.  This is correct because we don't visit blocks born in min_txg (only those born after -- see line 260), so it doesn't matter if the new feature was in effect for all, some, or none of the writes that happened in that txg.